tls/s2n_record_write.c

```diff
@@ -64,49 +64,51 @@ static int s2n_tls_record_overhead(struct s2n_connection *conn)
     return extra;
 }
 
-int s2n_record_rounded_write_payload_size(struct s2n_connection *conn, uint16_t size_without_overhead)
+/* This function returns maximum size of plaintext data to write for the payload.
+ * Record overheads are not included here.
+ */
+int s2n_record_max_write_payload_size(struct s2n_connection *conn)
 {
-    const struct s2n_crypto_parameters *active = conn->mode == S2N_CLIENT ? conn->client : conn->server;
+    notnull_check(conn);
+
+    const uint16_t max_fragment_size = MIN(conn->max_outgoing_fragment_length, S2N_TLS_MAXIMUM_FRAGMENT_LENGTH);
+
+    ENSURE_POSIX(max_fragment_size > 0, S2N_ERR_FRAGMENT_LENGTH_TOO_SMALL);
+    ENSURE_POSIX(max_fragment_size <= S2N_TLS_MAXIMUM_FRAGMENT_LENGTH, S2N_ERR_FRAGMENT_LENGTH_TOO_LARGE);
+
+    return max_fragment_size;
+}
+
+/* Find the largest size that will fit within an ethernet frame for a "small" payload */
+int s2n_record_min_write_payload_size(struct s2n_connection *conn)
+{
+    /* remove ethernet, TCP/IP and TLS header overheads */
+    const uint16_t min_outgoing_fragment_length = ETH_MTU - (conn->ipv6 ? IP_V6_HEADER_LENGTH : IP_V4_HEADER_LENGTH)
+        - TCP_HEADER_LENGTH - TCP_OPTIONS_LENGTH - S2N_TLS_RECORD_HEADER_LENGTH;
+
+    ENSURE_POSIX(min_outgoing_fragment_length <= S2N_TLS_MAXIMUM_FRAGMENT_LENGTH, S2N_ERR_FRAGMENT_LENGTH_TOO_LARGE);
+    int size = min_outgoing_fragment_length;
+
+    /* subtract overheads of a TLS record */
+    int overhead;
+    GUARD(overhead = s2n_tls_record_overhead(conn));
+    size -= overhead;
```

Comment:
why are we mixing `int` and `uint16_t` types?  can we use output parameters?  and can we `ENSURE(size > overhead);`?